File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if head is None:
            return head
            
        stack = []
        current = head

        while current:
            stack.append(current.val)
            current = current.next

        new_head = ListNode(stack.pop())
        temp = new_head
        i = len(stack) - 1

        while i>=0:
            next_el = ListNode(stack[i])
            temp.next = next_el
            temp = temp.next
            i-=1

        return new_head

File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        stack<int> st;

        if (head == nullptr || head->next == nullptr) return head;

        ListNode* temp = head;

        while(temp!=nullptr){
            st.push(temp->val);
            temp = temp->next;
        }

        ListNode* newHead = new ListNode(st.top());
        st.pop();
        ListNode* current = newHead;
        while(!st.empty()){
            ListNode* temp1 = new ListNode(st.top());
            st.pop();
            current->next = temp1;
            current = current->next;
        }

        return newHead;
    }
};